rllib/algorithms/algorithm.py

```diff
@@ -1774,17 +1890,74 @@ def import_policy_model_from_h5(
 
     @override(Trainable)
     def save_checkpoint(self, checkpoint_dir: str) -> str:
-        checkpoint_path = os.path.join(
-            checkpoint_dir, "checkpoint-{}".format(self.iteration)
-        )
-        pickle.dump(self.__getstate__(), open(checkpoint_path, "wb"))
+        """Exports AIR Checkpoint to a local directory and returns its directory path.
+
+        The structure of an Algorithm checkpoint dir will be as follows::
+
+            policies/
+                pol_1/
+                    policy_state.pkl
+                pol_2/
+                    policy_state.pkl
+            checkpoint_version.txt
+            state.pkl
 
-        return checkpoint_path
+        Note: `checkpoint_version.txt` contains a version string (e.g. "v0") helping
+        RLlib to remain backward compatible wrt restoring from checkpoints from
+        Ray 2.0 onwards.
+
+        Args:
+            checkpoint_dir: The directory where the checkpoint files will be stored.
+
+        Returns:
+            The path to the created AIR Checkpoint directory.
+        """
+        state = self.__getstate__()
+
+        # Extract policy states from worker state (Policies get their own
+        # checkpoint sub-dirs).
+        # TODO: "state" key inside "worker" should be renamed to "policy_states".
+        policies = {}
+        if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            policies = state["worker"].pop("state", {})
+
+        # Add RLlib checkpoint version.
+        state["__version__"] = "v1"
+
+        # Write state (w/o policies) to disk.
+        state_file = os.path.join(checkpoint_dir, "state.pkl")
+        pickle.dump(state, open(state_file, "wb"))
+        # Write checkpoint version separately. This file will NOT be used
+        # by RLlib anywhere, it is solely for the user's convenience.
+        with open(os.path.join(checkpoint_dir, "checkpoint_version.txt"), "w") as f:
+            f.write(state["__version__"])
+
+        # Write individual policies to disk.
+        if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            for pid, policy in self.workers.local_worker().policy_map.items():
+                policy_dir = os.path.join(checkpoint_dir, "policies", pid)
```

Comment:
this is not safe. pid can be an arbitrary string that contains illegal chars as dir name.
we need to sanitize it.
I don't know if Ray already has a util somewhere. if not, we can probably create one.

Reply:
Yes, still dangerous, though. What if a user choses:
```
multiagent = {
   policies={
      "pol:1": PolicySpec(),
      "pol_1": PolicySpec(),
      "pol1": PolicySpec(),
   }
}
```

:D

We should disallow policyIDs that contain any characters matching `\W`.

Reply:
wait really? are we really banning ``\W`` now?
I feel like ``pol_1`` looks totally legit? also I am worried about slashes and back-slashes, like "random_policy/policy1".
I don't think we rely on the dir name right? the actual policy_id is saved with the state dict.
we just need to make sure we construct a safe dir name, like replace everything `\W` with `_`.

Reply:
`\W` does not(!) include "_":

`\w == [a-zA-Z_]`

If we don't ban characters that are not allowed in dir names, this whole thing will not work. We don't have to be as strict as `\w`, we can allow anything that doesn't offend the OS for creating dir names, like:
`[^<>:"/\\|?*]` (not any of <<-- these characters, but all others are fine).

Slashes (backward or forward) are definitely a no-go, though!